delocate/tests/test_libsana.py

```diff
@@ -182,5 +183,5 @@ def test_resolve_rpath():
     lib_rpath = pjoin('@rpath', lib)
     # Should skip '/nonexist' path
     assert_equal(resolve_rpath(lib_rpath, ['/nonexist', path]), realpath(LIBA))
-    # Should return the given parameter as is since it can't be found
-    assert_equal(resolve_rpath(lib_rpath, []), lib_rpath)
+    # Should raise DependencyNotFound if the dependency can not be resolved.
```

Comment:
And test for warning, if that seems useful?

Reply:
The skip missing behavior is in the `tree_libs` function.  I changed the feedback from using the warnings module to just print, since it'd be expected behavior by that point.

Reply:
I have it back to using warnings now.

Reply:
@HexDecimal 
I checked out this branch to compare the PR code with the released `0.8.2`, after hitting this output when running `delocate-listdeps`. I have to say, it doesn't feel particularly right to me, in either the previous or current form. (Though it's definitely _much_ improved, now that it's no longer outputting module source code as the warning message!)

Even with the fixed code, though, it doesn't feel like the right sort of output _when running an interactive utility_. (Meaning, in the context of `delocate-listdeps` specifically.) In that context, in fact, it feels a bit like a violation of the ["When to use logging"](https://docs.python.org/3/howto/logging.html#when-to-use-logging) chart from the Basic logging tutorial, which advises:

|Task you want to perform                       | The best tool for the task  |
|------------------------------------------------|----------------------------------|
| Display console output for ordinary usage of a command line script or program | `print()` |
| Issue a warning regarding a particular runtime event | `warnings.warn()` in library code if the issue is avoidable and the client application should be modified to eliminate the warning |

...The primary implication here being that the warning signals an _application_ error, that "the client application should be modified to eliminate the warning". There's no change that can be made to `delocate-listdeps` to correct unresolvable dependency paths in a native binary. Shouldn't that just be reported in the normal `print()` output, like the rest of the dependency listing?

Reply:
I've been having the same feeling and I wish the logging module itself was being used.  There's a lot of stuff I wished I could print for debugging without it being in the main output.

Would it be valid to import the logging module and start using it for this kind of thing?  Or should I just stick with print?

Reply:
It's a bit tricky, because of the usual tension between libraries and command line utilities.   Printing from libraries is usually very bad style, and delocate can be used as a library.   So, what I'd like to find is some way of making the output behave as expected for a command line utilty - with ordinary, expected output going to stdout, and warnings / errors going to stderr - while also making it behave sensibly when used as a library.

Reply:
I think, precisely because of that multi-purpose utility, **no** decision about how to respond to a missing dependency can really be made at that low a level -- not to the point of informing the user, at least. Not by any method. 

Because, if the caller id `delocate-listdeps`, then an unresolvable path is merely a point of information, no more or less bad than a resolved path, and the tool should handle informing the user of both through normal means. There's no error of any kind, in that situation.

But if the caller is `delocate-wheel` trying to fix up package's external dependencies, not being able to resolve one is a four-alarm failure condition. So **then** it needs to be shouted about, in a way that user can't possibly miss.

That's a decision that can only be made higher up the call chain, with benefit of the proper context to make the right call. **Any** action taken way down in the inner workings of the library _other_ than kicking the problem upstairs will invariably be the wrong one, for some subset of all the possible scenarios.